client/client_test.go

```diff
@@ -182,3 +209,81 @@ func (s *ClientSuite) TestNewTargets(c *C) {
 // * Test locally expired metadata is ok
 // * Test invalid timestamp / snapshot signature downloads root.json
 // * Test invalid hash returns ErrDownloadFailed
+
+type testDestination struct {
+	bytes.Buffer
+	deleted bool
+}
+
+func (t *testDestination) Delete() error {
+	t.deleted = true
+	return nil
+}
+
+func (s *ClientSuite) TestDownloadNoExist(c *C) {
+	client := s.updatedClient(c)
+	var dest testDestination
+	c.Assert(client.Download("nonexistent", &dest), Equals, ErrNotFound{"nonexistent"})
+	c.Assert(dest.deleted, Equals, true)
+}
+
+func (s *ClientSuite) TestDownloadOK(c *C) {
+	client := s.updatedClient(c)
+	var dest testDestination
+	c.Assert(client.Download("foo.txt", &dest), IsNil)
+	c.Assert(dest.deleted, Equals, false)
+	c.Assert(dest.String(), Equals, "foo")
+}
+
+func (s *ClientSuite) TestDownloadTargetTooLong(c *C) {
+	client := s.updatedClient(c)
+	remoteFile := s.remote["targets/foo.txt"]
+	remoteFile.buf = bytes.NewReader([]byte("foo-ooo"))
+	var dest testDestination
+	c.Assert(client.Download("foo.txt", &dest), IsNil)
+	c.Assert(remoteFile.bytesRead, Equals, 3)
+	c.Assert(dest.deleted, Equals, false)
```

Comment:
It's hard to think of a scenario where something that odd happens and it's still correct, so I'd agree with rejecting that state.